File: branches/V1.0/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/reports/DocumentAccountIndex.java
package com.jasonzqshen.familyaccounting.core.reports;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataIdentity;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataManagement;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;
import com.jasonzqshen.familyaccounting.core.transaction.ItemEntity;
import com.jasonzqshen.familyaccounting.core.utils.CreditDebitIndicator;
import com.jasonzqshen.familyaccounting.core.utils.CurrencyAmount;

public class DocumentAccountIndex extends DocumentIndex {

	DocumentAccountIndex(CoreDriver coreDriver, MasterDataManagement mdMgmt) {
		super(coreDriver, mdMgmt);
	}

	@Override
	protected void newDoc(HeadEntity head) {
		ItemEntity[] items = head.getItems();

		for (int i = 0; i < items.length; ++i) {
			MasterDataIdentity id = items[i].getGLAccount();
			DocumentIndexItem item;

			if (!_list.containsKey(id)) {
				item = new DocumentIndexItem(_coreDriver, id);
				_list.put(id, item);
			} else {
				item = _list.get(id);
			}
			item.addDoc(head);
			// add amount
			CurrencyAmount amount = items[i].getAmount();
			if (items[i].getCDIndicator() == CreditDebitIndicator.CREDIT) {
				amount.negate();
			}
			item.addAmount(head.getMonthId(), amount);
		}
	}

	@Override
	protected void reverseDoc(HeadEntity head) {
		ItemEntity[] items = head.getItems();

		for (int i = 0; i < items.length; ++i) {
			MasterDataIdentity id = items[i].getGLAccount();
			if (!_list.containsKey(id)) {
				continue;
			}
			DocumentIndexItem item = _list.get(id);

			// remove document
			item.removeDoc(head);
			// remove amount
			CurrencyAmount amount = items[i].getAmount();
			if (items[i].getCDIndicator() == CreditDebitIndicator.DEBIT) {
				amount.negate();
			}
			item.addAmount(head.getMonthId(), amount);
		}
	}
}

File: branches/V1.0/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Reports/DocumentAccountIndex.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using WinStore_FamilyAccountingCore.Utilities;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;

namespace WinStore_FamilyAccountingCore.Reports
{
    /// <summary>
    /// Account index
    /// </summary>
    public class DocumentAccountIndex : DocumentIndex
    {

        internal DocumentAccountIndex(CoreDriver coreDriver, MasterDataManagement mdMgmt)
            : base(coreDriver, mdMgmt)
        {
        }

        /// <summary>
        /// new document
        /// </summary>
        /// <param name="head"></param>
        protected override void newDoc(HeadEntity head)
        {
            List<ItemEntity> items = head.Items;

            foreach (ItemEntity itemEntity in items)
            {
                MasterDataIdentity id = itemEntity.GLAccount;
                DocumentIndexItem item;

                if (!_list.TryGetValue(id, out item))
                {
                    item = new DocumentIndexItem(_coreDriver, id);
                    _list.Add(id, item);
                }
                item.addDoc(head);
                // add amount
                CurrencyAmount amount = itemEntity.Amount;
                if (itemEntity.CdIndicator == CreditDebitIndicator.CREDIT)
                {
                    amount.Negate();
                }
                item.addAmount(head.MonthID, amount);
            }
        }
        /// <summary>
        /// reverse document
        /// </summary>
        /// <param name="head"></param>
        protected override void reverseDoc(HeadEntity head)
        {
            List<ItemEntity> items = head.Items;

            foreach (ItemEntity itemEntity in items)
            {
                MasterDataIdentity id = itemEntity.GLAccount;
                if (!_list.ContainsKey(id))
                {
                    continue;
                }
                DocumentIndexItem item;
                if (_list.TryGetValue(id, out item) == false)
                {
                    continue;
                }

                // remove document
                item.removeDoc(head);
                // remove amount
                CurrencyAmount amount = itemEntity.Amount;
                if (itemEntity.CdIndicator == CreditDebitIndicator.DEBIT)
                {
                    amount.Negate();
                }
                item.addAmount(head.MonthID, amount);
            }
        }
    }

}
